src/mewbot/bot.py

```diff
@@ -68,46 +84,75 @@ class BotRunner:
     input_event_queue: InputQueue
     output_event_queue: OutputQueue
 
-    outputs: Dict[Type[OutputEvent], Set[Output]] = {}
+    inputs: Set[InputInterface]
+    outputs: Dict[Type[OutputEvent], Set[OutputInterface]] = {}
     behaviours: Dict[Type[InputEvent], Set[Behaviour]] = {}
 
     def __init__(
         self,
         behaviours: Dict[Type[InputEvent], Set[Behaviour]],
-        inputs: Set[Input],
-        outputs: Dict[Type[OutputEvent], Set[Output]],
+        inputs: Set[InputInterface],
+        outputs: Dict[Type[OutputEvent], Set[OutputInterface]],
     ) -> None:
+
+        self.logger = logging.getLogger(__name__ + "BotRunner")
+
         self.input_event_queue = InputQueue()
         self.output_event_queue = OutputQueue()
+
         self.inputs = inputs
         self.outputs = outputs
         self.behaviours = behaviours
 
-    def run(self) -> None:
+    def run(self, loop: Optional[asyncio.AbstractEventLoop] = None) -> None:
+        loop = loop if loop else asyncio.get_event_loop()
+
+        # Handle correctly terminating the loop
+        try:
+            loop.add_signal_handler(signal.SIGINT, loop.stop)
+            loop.add_signal_handler(signal.SIGTERM, loop.stop)
+        except NotImplementedError:
+            pass
+
+        loop.create_task(self.create_tasks(loop))
+        try:
+            loop.run_forever()
+        finally:
+            loop.close()
+
+        # Seemed to fix a shutdown bug - now less sure
+        asyncio.ensure_future(self.create_tasks(loop), loop=loop)
+        # Need a running event loop for some of the io init tasks
+
+    async def create_tasks(self, loop: asyncio.AbstractEventLoop) -> None:
+        self.logger.info("Starting main event loop")
+
+        # Startup the inputs
+        self.logger.info("self.inputs: %s", ",".join(map(str, self.inputs)))
```

Comment:
Not checked, but want to note a moment of care may need to happen here if the stringification of self.inputs captures the configuration of each input and any auth tokens if they are bound up in that. (which then gets logged presumably...)

Reply:
By the spec, they should not just be dumping dicts of their own state (https://github.com/mewler/mewbot/tree/mainline/design-docs#human-readable-descriptions), but the burden of defining sensible str representation of components is up to the people implementing that component.